File: apps/mobile/shared/Sources/FlagshipCore/DaemonStatus.swift
import Foundation
import CryptoKit
import FlagshipAPI

/// Swift mirror of `packages/protocol/src/daemonStatus.ts` — canonical bytes
/// + Ed25519 verify for the box's STK-signed daemon-status report, the
/// cert-fingerprint pinning primitive (cert-model A′, phase 4).
///
/// The box signs the tuple with its STK; `.com` relays it VERBATIM on
/// `/pods` (`signedStatus: { report, signatureHex }`). A phone that derives
/// the box STK locally (`ServerKeys.deriveStkPub` — NOT `.com`'s
/// `identityPubKey` echo) re-verifies the leaf-cert fingerprint end-to-end,
/// so a rogue `.com` can DROP a report but cannot FORGE one.
///
/// Canonical bytes MUST match the TS implementation byte-for-byte (pinned
/// cross-platform vector in packages/protocol/tests/daemonStatus.test.ts +
/// DaemonStatusVerifierTests here):
///
///   flagship/daemon-status/v1|<serverDomain>|<certSha256 or "">|
///   <certValidUntil or "">|<certIssuer or "">|<appsServed sorted, ","-joined>|
///   <nonce>|<issuedAt>
public enum DaemonStatus {
    public static let canonicalTag = "flagship/daemon-status/v1"

    /// Freshness bound on a relayed report: older than this and the
    /// fingerprint is NOT pinned. The daemon heartbeat is 5-minutely, so 7
    /// days is generous — a report this stale means the box (or the relay)
    /// has been silent long enough that its cert may have legitimately
    /// renewed; falling back to default TLS validation (no pin) is the safe
    /// failure, never a hard fail.
    public static let maxReportAgeMs: Int64 = 7 * 24 * 60 * 60 * 1000

    public static func canonicalBytes(_ r: DaemonStatusReport) -> Data {
        let segments = [
            canonicalTag,
            r.serverDomain,
            r.certSha256 ?? "",
            r.certValidUntil.map(String.init) ?? "",
            r.certIssuer ?? "",
            r.appsServed.sorted().joined(separator: ","),
            r.nonce,
            String(r.issuedAt)
        ]
        return Data(segments.joined(separator: "|").utf8)
    }

    /// Verify the box's signature under its (locally derived) STK pubkey.
    /// Returns false — never throws — on malformed input, mirroring the TS
    /// `verifyDaemonStatusReport`.
    public static func verify(_ r: DaemonStatusReport, signature: Data, stkPub: Data) -> Bool {
        guard signature.count == 64,
              let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: stkPub)
        else { return false }
        return pub.isValidSignature(signature, for: canonicalBytes(r))
    }

    public static func verify(_ r: DaemonStatusReport, signatureHex: String, stkPub: Data) -> Bool {
        guard let sig = HexUtil.decode(signatureHex) else { return false }
        return verify(r, signature: sig, stkPub: stkPub)
    }
}

File: packages/protocol/src/daemonStatus.ts
/**
 * STK-signed daemon-status report — the cert-fingerprint pinning primitive
 * (cert-model A′, phase 4).
 *
 * The daemon periodically POSTs this report to .com; .com verifies it,
 * stores the VERBATIM signed tuple + signature, and relays both on /pods.
 * A phone that derived the box STK locally (deriveSTK(deriveSWK(UMK,
 * serverId))) re-verifies the leaf-cert fingerprint end-to-end, so a rogue
 * .com can DROP the report but cannot FORGE one — the fingerprint a client
 * pins is the box's own word, not .com's.
 *
 * Canonical bytes (one implementation, shared by the daemon heartbeat, the
 * control-plane verifier, and every client; iOS/Android mirror this
 * byte-for-byte — see the pinned vector in tests/daemonStatus.test.ts):
 *
 *   flagship/daemon-status/v1|<serverDomain>|<certSha256 or "">|
 *   <certValidUntil or "">|<certIssuer or "">|<appsServed sorted, ","-joined>|
 *   <nonce>|<issuedAt>
 */
import { ed } from "./edSync.js";
import { legacyFieldGuard } from "./auth.js";
import { resolveMsgSigner, type MsgSigner } from "./canonicalBase.js";
import type { Bytes } from "./types.js";

export interface DaemonStatusReport {
  serverDomain: string;
  /** Leaf-cert SHA-256 fingerprint: lowercase hex, no colons. Null when the
   *  box has no cert yet (liveness-only report). */
  certSha256: string | null;
  certValidUntil: number | null;
  certIssuer: string | null;
  appsServed: string[];
  nonce: string;
  issuedAt: number;
}

const TAG_DAEMON_STATUS = "flagship/daemon-status/v1";

export function canonicalDaemonStatusReport(r: DaemonStatusReport): Bytes {
  legacyFieldGuard("serverDomain", r.serverDomain);
  if (r.certSha256 !== null) legacyFieldGuard("certSha256", r.certSha256);
  if (r.certIssuer !== null) legacyFieldGuard("certIssuer", r.certIssuer);
  for (const app of r.appsServed) legacyFieldGuard("appServed", app);
  legacyFieldGuard("nonce", r.nonce);
  const apps = r.appsServed.slice().sort().join(",");
  return new TextEncoder().encode(
    [
      TAG_DAEMON_STATUS,
      r.serverDomain,
      r.certSha256 ?? "",
      String(r.certValidUntil ?? ""),
      r.certIssuer ?? "",
      apps,
      r.nonce,
      String(r.issuedAt),
    ].join("|"),
  );
}

export function signDaemonStatusReport(
  r: DaemonStatusReport,
  identity: MsgSigner,
): Bytes {
  // Accepts the STK `Keypair` or a `sign(msg)` closure (custodian-backed);
  // signature bytes are identical either way.
  return resolveMsgSigner(identity)(canonicalDaemonStatusReport(r));
}

export function verifyDaemonStatusReport(
  r: DaemonStatusReport,
  sig: Bytes,
  stkPub: Bytes,
): boolean {
  try {
    return ed.verify(sig, canonicalDaemonStatusReport(r), stkPub);
  } catch {
    return false;
  }
}
